File: packages/rule-engine-php/tests/FixturesTest.php
<?php

declare(strict_types=1);

namespace Karnek\RuleEngine\Tests;

use Karnek\RuleEngine\Engine;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

final class FixturesTest extends TestCase
{
    private const CASES = __DIR__ . '/../../../fixtures/rule-engine/cases';
    private const PACKS = __DIR__ . '/../../../packs';

    /** @return iterable<string, array{0: string}> */
    public static function cases(): iterable
    {
        foreach (glob(self::CASES . '/*.json') ?: [] as $path) {
            yield basename($path, '.json') => [$path];
        }
    }

    #[DataProvider('cases')]
    public function testCaseMatchesFixture(string $path): void
    {
        $case = json_decode((string) file_get_contents($path), true, 512, JSON_THROW_ON_ERROR);
        $engine = new Engine(static fn (string $firmId): array => json_decode(
            (string) file_get_contents(self::PACKS . '/' . $firmId . '.json'),
            true,
            512,
            JSON_THROW_ON_ERROR,
        ));

        if (isset($case['inputBefore'])) {
            $before = $engine->computeWindows($case['inputBefore']);
            $after = $engine->computeWindows($case['inputAfter']);
            $got = Engine::reconcile(
                $engine->computeLadder($case['inputBefore']['userId'], $before['windows']),
                $engine->computeLadder($case['inputAfter']['userId'], $after['windows']),
            );
            $got['windowsAfter'] = $after['windows'];
        } else {
            $result = $engine->computeWindows($case['input']);
            $got = [
                'windows' => $result['windows'],
                'ladder' => $engine->computeLadder($case['input']['userId'], $result['windows']),
                'notes' => $result['notes'],
            ];
        }

        self::assertSame(
            self::canonical($case['expected']),
            self::canonical($got),
            'Engine output drifted from ' . basename($path) . '. reference.py is the tie-breaker.',
        );
    }

    public function testEveryCaseHasAnExpectedBlock(): void
    {
        $count = 0;
        foreach (self::cases() as [$path]) {
            $case = json_decode((string) file_get_contents($path), true, 512, JSON_THROW_ON_ERROR);
            self::assertArrayHasKey('expected', $case, basename($path) . ' has no expected block; run reference.py --write');
            $count++;
        }
        self::assertSame(13, $count, 'expected thirteen fixture cases');
    }

    /** Sort associative keys recursively so key order never causes a false mismatch. */
    private static function canonical(mixed $value): mixed
    {
        if (!is_array($value)) {
            return $value;
        }
        $isList = array_is_list($value);
        $out = [];
        foreach ($value as $k => $v) {
            $out[$k] = self::canonical($v);
        }
        if (!$isList) {
            ksort($out);
        }

        return $out;
    }
}

File: packages/rule_engine/test/fixtures_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:rule_engine/rule_engine.dart';
import 'package:test/test.dart';

const casesDir = '../../fixtures/rule-engine/cases';
const packsDir = '../../packs';

Map<String, dynamic> loadPack(String firmId) =>
    jsonDecode(File('$packsDir/$firmId.json').readAsStringSync()) as Map<String, dynamic>;

/// Round-trips through JSON so Dart objects compare as plain maps and lists.
dynamic plain(Object? v) => jsonDecode(jsonEncode(v));

void main() {
  final files = Directory(casesDir)
      .listSync()
      .whereType<File>()
      .where((f) => f.path.endsWith('.json'))
      .toList()
    ..sort((a, b) => a.path.compareTo(b.path));

  test('there are thirteen fixture cases', () {
    expect(files.length, 13);
  });

  for (final file in files) {
    final name = file.uri.pathSegments.last;
    test(name, () {
      final json = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
      expect(json.containsKey('expected'), isTrue,
          reason: '$name has no expected block; run reference.py --write');
      final engine = RuleEngine(loadPack);

      final Map<String, dynamic> got;
      if (json.containsKey('inputBefore')) {
        final before = engine.computeWindows(json['inputBefore'] as Map<String, dynamic>);
        final after = engine.computeWindows(json['inputAfter'] as Map<String, dynamic>);
        final r = RuleEngine.reconcile(
          engine.computeLadder((json['inputBefore'] as Map)['userId'] as String, before.windows),
          engine.computeLadder((json['inputAfter'] as Map)['userId'] as String, after.windows),
        );
        got = {
          'cancelledAlertIds': r.cancelledAlertIds,
          'createdAlertIds': r.createdAlertIds,
          'keptAlertIds': r.keptAlertIds,
          'windowsAfter': after.windows.map((w) => w.toJson()).toList(),
        };
      } else {
        final input = json['input'] as Map<String, dynamic>;
        final result = engine.computeWindows(input);
        got = {
          'windows': result.windows.map((w) => w.toJson()).toList(),
          'ladder': engine
              .computeLadder(input['userId'] as String, result.windows)
              .map((r) => r.toJson())
              .toList(),
          'notes': result.notes,
        };
      }

      expect(plain(got), equals(plain(json['expected'])),
          reason: 'Engine output drifted from $name. reference.py is the tie-breaker.');
    });
  }
}
